// @wormhole-foundation/sdk-base/dist/cjs/utils/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterIndexes = exports.pickWithOrder = exports.zip = exports.column = exports.flatten = exports.entries = exports.range = void 0;
const range = (length) => [...Array(length).keys()];
exports.range = range;
function entries(arr) {
    return [...arr.entries()];
}
exports.entries = entries;
const flatten = (arr) => arr.flat();
exports.flatten = flatten;
const column = (tupArr, index) => tupArr.map((tuple) => tuple[index]);
exports.column = column;
const zip = (arr) => (0, exports.range)(arr[0].length).map(col => (0, exports.range)(arr.length).map(row => arr[row][col]));
exports.zip = zip;
const pickWithOrder = (arr, indexes) => indexes.map((i) => arr[i]);
exports.pickWithOrder = pickWithOrder;
const filterIndexes = (arr, indexes, exclude) => {
    const indexSet = new Set(Array.isArray(indexes) ? indexes : [indexes]);
    return arr.filter((_, i) => indexSet.has(i) !== exclude);
};
exports.filterIndexes = filterIndexes;
