// @injectivelabs/sdk-ts/dist/cjs/core/modules/permissions/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgUpdateParams = exports.MsgUpdateNamespaceRoles = exports.MsgUpdateNamespace = exports.MsgRevokeNamespaceRoles = exports.MsgDeleteNamespace = exports.MsgCreateNamespace = exports.MsgClaimVoucher = void 0;
const MsgClaimVoucher_js_1 = __importDefault(require("./msgs/MsgClaimVoucher.js"));
exports.MsgClaimVoucher = MsgClaimVoucher_js_1.default;
const MsgCreateNamespace_js_1 = __importDefault(require("./msgs/MsgCreateNamespace.js"));
exports.MsgCreateNamespace = MsgCreateNamespace_js_1.default;
const MsgDeleteNamespace_js_1 = __importDefault(require("./msgs/MsgDeleteNamespace.js"));
exports.MsgDeleteNamespace = MsgDeleteNamespace_js_1.default;
const MsgRevokeNamespaceRoles_js_1 = __importDefault(require("./msgs/MsgRevokeNamespaceRoles.js"));
exports.MsgRevokeNamespaceRoles = MsgRevokeNamespaceRoles_js_1.default;
const MsgUpdateNamespace_js_1 = __importDefault(require("./msgs/MsgUpdateNamespace.js"));
exports.MsgUpdateNamespace = MsgUpdateNamespace_js_1.default;
const MsgUpdateNamespaceRoles_js_1 = __importDefault(require("./msgs/MsgUpdateNamespaceRoles.js"));
exports.MsgUpdateNamespaceRoles = MsgUpdateNamespaceRoles_js_1.default;
const MsgUpdateParams_js_1 = __importDefault(require("./msgs/MsgUpdateParams.js"));
exports.MsgUpdateParams = MsgUpdateParams_js_1.default;
