// @injectivelabs/core-proto-ts/cjs/injective/permissions/v1beta1/tx.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcWebError = exports.GrpcWebImpl = exports.MsgClaimVoucherDesc = exports.MsgRevokeNamespaceRolesDesc = exports.MsgUpdateNamespaceRolesDesc = exports.MsgUpdateNamespaceDesc = exports.MsgDeleteNamespaceDesc = exports.MsgCreateNamespaceDesc = exports.MsgUpdateParamsDesc = exports.MsgDesc = exports.MsgClientImpl = exports.MsgClaimVoucherResponse = exports.MsgClaimVoucher = exports.MsgRevokeNamespaceRolesResponse = exports.MsgRevokeNamespaceRoles = exports.MsgUpdateNamespaceRolesResponse = exports.MsgUpdateNamespaceRoles = exports.MsgUpdateNamespaceResponse = exports.MsgUpdateNamespace_MsgSetBurnsPaused = exports.MsgUpdateNamespace_MsgSetSendsPaused = exports.MsgUpdateNamespace_MsgSetMintsPaused = exports.MsgUpdateNamespace_MsgSetWasmHook = exports.MsgUpdateNamespace = exports.MsgDeleteNamespaceResponse = exports.MsgDeleteNamespace = exports.MsgCreateNamespaceResponse = exports.MsgCreateNamespace = exports.MsgUpdateParamsResponse = exports.MsgUpdateParams = exports.protobufPackage = void 0;
/* eslint-disable */
var grpc_web_1 = require("@injectivelabs/grpc-web");
var browser_headers_1 = require("browser-headers");
var minimal_js_1 = __importDefault(require("protobufjs/minimal.js"));
var params_1 = require("./params.js");
var permissions_1 = require("./permissions.js");
exports.protobufPackage = "injective.permissions.v1beta1";
function createBaseMsgUpdateParams() {
    return { authority: "", params: undefined };
}
exports.MsgUpdateParams = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.authority !== "") {
            writer.uint32(10).string(message.authority);
        }
        if (message.params !== undefined) {
            params_1.Params.encode(message.params, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateParams();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.authority = reader.string();
                    break;
                case 2:
                    message.params = params_1.Params.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            authority: isSet(object.authority) ? String(object.authority) : "",
            params: isSet(object.params) ? params_1.Params.fromJSON(object.params) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.authority !== undefined && (obj.authority = message.authority);
        message.params !== undefined && (obj.params = message.params ? params_1.Params.toJSON(message.params) : undefined);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateParams.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgUpdateParams();
        message.authority = (_a = object.authority) !== null && _a !== void 0 ? _a : "";
        message.params = (object.params !== undefined && object.params !== null)
            ? params_1.Params.fromPartial(object.params)
            : undefined;
        return message;
    },
};
function createBaseMsgUpdateParamsResponse() {
    return {};
}
exports.MsgUpdateParamsResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateParamsResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateParamsResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgUpdateParamsResponse();
        return message;
    },
};
function createBaseMsgCreateNamespace() {
    return { sender: "", namespace: undefined };
}
exports.MsgCreateNamespace = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.namespace !== undefined) {
            permissions_1.Namespace.encode(message.namespace, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgCreateNamespace();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.namespace = permissions_1.Namespace.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            namespace: isSet(object.namespace) ? permissions_1.Namespace.fromJSON(object.namespace) : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.namespace !== undefined &&
            (obj.namespace = message.namespace ? permissions_1.Namespace.toJSON(message.namespace) : undefined);
        return obj;
    },
    create: function (base) {
        return exports.MsgCreateNamespace.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgCreateNamespace();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.namespace = (object.namespace !== undefined && object.namespace !== null)
            ? permissions_1.Namespace.fromPartial(object.namespace)
            : undefined;
        return message;
    },
};
function createBaseMsgCreateNamespaceResponse() {
    return {};
}
exports.MsgCreateNamespaceResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgCreateNamespaceResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgCreateNamespaceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgCreateNamespaceResponse();
        return message;
    },
};
function createBaseMsgDeleteNamespace() {
    return { sender: "", namespaceDenom: "" };
}
exports.MsgDeleteNamespace = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.namespaceDenom !== "") {
            writer.uint32(18).string(message.namespaceDenom);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgDeleteNamespace();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.namespaceDenom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            namespaceDenom: isSet(object.namespaceDenom) ? String(object.namespaceDenom) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.namespaceDenom !== undefined && (obj.namespaceDenom = message.namespaceDenom);
        return obj;
    },
    create: function (base) {
        return exports.MsgDeleteNamespace.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseMsgDeleteNamespace();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.namespaceDenom = (_b = object.namespaceDenom) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseMsgDeleteNamespaceResponse() {
    return {};
}
exports.MsgDeleteNamespaceResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgDeleteNamespaceResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgDeleteNamespaceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgDeleteNamespaceResponse();
        return message;
    },
};
function createBaseMsgUpdateNamespace() {
    return {
        sender: "",
        namespaceDenom: "",
        wasmHook: undefined,
        mintsPaused: undefined,
        sendsPaused: undefined,
        burnsPaused: undefined,
    };
}
exports.MsgUpdateNamespace = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.namespaceDenom !== "") {
            writer.uint32(18).string(message.namespaceDenom);
        }
        if (message.wasmHook !== undefined) {
            exports.MsgUpdateNamespace_MsgSetWasmHook.encode(message.wasmHook, writer.uint32(26).fork()).ldelim();
        }
        if (message.mintsPaused !== undefined) {
            exports.MsgUpdateNamespace_MsgSetMintsPaused.encode(message.mintsPaused, writer.uint32(34).fork()).ldelim();
        }
        if (message.sendsPaused !== undefined) {
            exports.MsgUpdateNamespace_MsgSetSendsPaused.encode(message.sendsPaused, writer.uint32(42).fork()).ldelim();
        }
        if (message.burnsPaused !== undefined) {
            exports.MsgUpdateNamespace_MsgSetBurnsPaused.encode(message.burnsPaused, writer.uint32(50).fork()).ldelim();
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespace();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.namespaceDenom = reader.string();
                    break;
                case 3:
                    message.wasmHook = exports.MsgUpdateNamespace_MsgSetWasmHook.decode(reader, reader.uint32());
                    break;
                case 4:
                    message.mintsPaused = exports.MsgUpdateNamespace_MsgSetMintsPaused.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.sendsPaused = exports.MsgUpdateNamespace_MsgSetSendsPaused.decode(reader, reader.uint32());
                    break;
                case 6:
                    message.burnsPaused = exports.MsgUpdateNamespace_MsgSetBurnsPaused.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            namespaceDenom: isSet(object.namespaceDenom) ? String(object.namespaceDenom) : "",
            wasmHook: isSet(object.wasmHook) ? exports.MsgUpdateNamespace_MsgSetWasmHook.fromJSON(object.wasmHook) : undefined,
            mintsPaused: isSet(object.mintsPaused)
                ? exports.MsgUpdateNamespace_MsgSetMintsPaused.fromJSON(object.mintsPaused)
                : undefined,
            sendsPaused: isSet(object.sendsPaused)
                ? exports.MsgUpdateNamespace_MsgSetSendsPaused.fromJSON(object.sendsPaused)
                : undefined,
            burnsPaused: isSet(object.burnsPaused)
                ? exports.MsgUpdateNamespace_MsgSetBurnsPaused.fromJSON(object.burnsPaused)
                : undefined,
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.namespaceDenom !== undefined && (obj.namespaceDenom = message.namespaceDenom);
        message.wasmHook !== undefined &&
            (obj.wasmHook = message.wasmHook ? exports.MsgUpdateNamespace_MsgSetWasmHook.toJSON(message.wasmHook) : undefined);
        message.mintsPaused !== undefined && (obj.mintsPaused = message.mintsPaused
            ? exports.MsgUpdateNamespace_MsgSetMintsPaused.toJSON(message.mintsPaused)
            : undefined);
        message.sendsPaused !== undefined && (obj.sendsPaused = message.sendsPaused
            ? exports.MsgUpdateNamespace_MsgSetSendsPaused.toJSON(message.sendsPaused)
            : undefined);
        message.burnsPaused !== undefined && (obj.burnsPaused = message.burnsPaused
            ? exports.MsgUpdateNamespace_MsgSetBurnsPaused.toJSON(message.burnsPaused)
            : undefined);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespace.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseMsgUpdateNamespace();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.namespaceDenom = (_b = object.namespaceDenom) !== null && _b !== void 0 ? _b : "";
        message.wasmHook = (object.wasmHook !== undefined && object.wasmHook !== null)
            ? exports.MsgUpdateNamespace_MsgSetWasmHook.fromPartial(object.wasmHook)
            : undefined;
        message.mintsPaused = (object.mintsPaused !== undefined && object.mintsPaused !== null)
            ? exports.MsgUpdateNamespace_MsgSetMintsPaused.fromPartial(object.mintsPaused)
            : undefined;
        message.sendsPaused = (object.sendsPaused !== undefined && object.sendsPaused !== null)
            ? exports.MsgUpdateNamespace_MsgSetSendsPaused.fromPartial(object.sendsPaused)
            : undefined;
        message.burnsPaused = (object.burnsPaused !== undefined && object.burnsPaused !== null)
            ? exports.MsgUpdateNamespace_MsgSetBurnsPaused.fromPartial(object.burnsPaused)
            : undefined;
        return message;
    },
};
function createBaseMsgUpdateNamespace_MsgSetWasmHook() {
    return { newValue: "" };
}
exports.MsgUpdateNamespace_MsgSetWasmHook = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.newValue !== "") {
            writer.uint32(10).string(message.newValue);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespace_MsgSetWasmHook();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.newValue = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { newValue: isSet(object.newValue) ? String(object.newValue) : "" };
    },
    toJSON: function (message) {
        var obj = {};
        message.newValue !== undefined && (obj.newValue = message.newValue);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespace_MsgSetWasmHook.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgUpdateNamespace_MsgSetWasmHook();
        message.newValue = (_a = object.newValue) !== null && _a !== void 0 ? _a : "";
        return message;
    },
};
function createBaseMsgUpdateNamespace_MsgSetMintsPaused() {
    return { newValue: false };
}
exports.MsgUpdateNamespace_MsgSetMintsPaused = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.newValue === true) {
            writer.uint32(8).bool(message.newValue);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespace_MsgSetMintsPaused();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.newValue = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { newValue: isSet(object.newValue) ? Boolean(object.newValue) : false };
    },
    toJSON: function (message) {
        var obj = {};
        message.newValue !== undefined && (obj.newValue = message.newValue);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespace_MsgSetMintsPaused.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgUpdateNamespace_MsgSetMintsPaused();
        message.newValue = (_a = object.newValue) !== null && _a !== void 0 ? _a : false;
        return message;
    },
};
function createBaseMsgUpdateNamespace_MsgSetSendsPaused() {
    return { newValue: false };
}
exports.MsgUpdateNamespace_MsgSetSendsPaused = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.newValue === true) {
            writer.uint32(8).bool(message.newValue);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespace_MsgSetSendsPaused();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.newValue = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { newValue: isSet(object.newValue) ? Boolean(object.newValue) : false };
    },
    toJSON: function (message) {
        var obj = {};
        message.newValue !== undefined && (obj.newValue = message.newValue);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespace_MsgSetSendsPaused.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgUpdateNamespace_MsgSetSendsPaused();
        message.newValue = (_a = object.newValue) !== null && _a !== void 0 ? _a : false;
        return message;
    },
};
function createBaseMsgUpdateNamespace_MsgSetBurnsPaused() {
    return { newValue: false };
}
exports.MsgUpdateNamespace_MsgSetBurnsPaused = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.newValue === true) {
            writer.uint32(8).bool(message.newValue);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespace_MsgSetBurnsPaused();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.newValue = reader.bool();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return { newValue: isSet(object.newValue) ? Boolean(object.newValue) : false };
    },
    toJSON: function (message) {
        var obj = {};
        message.newValue !== undefined && (obj.newValue = message.newValue);
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespace_MsgSetBurnsPaused.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a;
        var message = createBaseMsgUpdateNamespace_MsgSetBurnsPaused();
        message.newValue = (_a = object.newValue) !== null && _a !== void 0 ? _a : false;
        return message;
    },
};
function createBaseMsgUpdateNamespaceResponse() {
    return {};
}
exports.MsgUpdateNamespaceResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespaceResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespaceResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgUpdateNamespaceResponse();
        return message;
    },
};
function createBaseMsgUpdateNamespaceRoles() {
    return { sender: "", namespaceDenom: "", rolePermissions: [], addressRoles: [] };
}
exports.MsgUpdateNamespaceRoles = {
    encode: function (message, writer) {
        var e_1, _a, e_2, _b;
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.namespaceDenom !== "") {
            writer.uint32(18).string(message.namespaceDenom);
        }
        try {
            for (var _c = __values(message.rolePermissions), _d = _c.next(); !_d.done; _d = _c.next()) {
                var v = _d.value;
                permissions_1.Role.encode(v, writer.uint32(26).fork()).ldelim();
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        try {
            for (var _e = __values(message.addressRoles), _f = _e.next(); !_f.done; _f = _e.next()) {
                var v = _f.value;
                permissions_1.AddressRoles.encode(v, writer.uint32(34).fork()).ldelim();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespaceRoles();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.namespaceDenom = reader.string();
                    break;
                case 3:
                    message.rolePermissions.push(permissions_1.Role.decode(reader, reader.uint32()));
                    break;
                case 4:
                    message.addressRoles.push(permissions_1.AddressRoles.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            namespaceDenom: isSet(object.namespaceDenom) ? String(object.namespaceDenom) : "",
            rolePermissions: Array.isArray(object === null || object === void 0 ? void 0 : object.rolePermissions)
                ? object.rolePermissions.map(function (e) { return permissions_1.Role.fromJSON(e); })
                : [],
            addressRoles: Array.isArray(object === null || object === void 0 ? void 0 : object.addressRoles)
                ? object.addressRoles.map(function (e) { return permissions_1.AddressRoles.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.namespaceDenom !== undefined && (obj.namespaceDenom = message.namespaceDenom);
        if (message.rolePermissions) {
            obj.rolePermissions = message.rolePermissions.map(function (e) { return e ? permissions_1.Role.toJSON(e) : undefined; });
        }
        else {
            obj.rolePermissions = [];
        }
        if (message.addressRoles) {
            obj.addressRoles = message.addressRoles.map(function (e) { return e ? permissions_1.AddressRoles.toJSON(e) : undefined; });
        }
        else {
            obj.addressRoles = [];
        }
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespaceRoles.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c, _d;
        var message = createBaseMsgUpdateNamespaceRoles();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.namespaceDenom = (_b = object.namespaceDenom) !== null && _b !== void 0 ? _b : "";
        message.rolePermissions = ((_c = object.rolePermissions) === null || _c === void 0 ? void 0 : _c.map(function (e) { return permissions_1.Role.fromPartial(e); })) || [];
        message.addressRoles = ((_d = object.addressRoles) === null || _d === void 0 ? void 0 : _d.map(function (e) { return permissions_1.AddressRoles.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseMsgUpdateNamespaceRolesResponse() {
    return {};
}
exports.MsgUpdateNamespaceRolesResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgUpdateNamespaceRolesResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgUpdateNamespaceRolesResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgUpdateNamespaceRolesResponse();
        return message;
    },
};
function createBaseMsgRevokeNamespaceRoles() {
    return { sender: "", namespaceDenom: "", addressRolesToRevoke: [] };
}
exports.MsgRevokeNamespaceRoles = {
    encode: function (message, writer) {
        var e_3, _a;
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.namespaceDenom !== "") {
            writer.uint32(18).string(message.namespaceDenom);
        }
        try {
            for (var _b = __values(message.addressRolesToRevoke), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                permissions_1.AddressRoles.encode(v, writer.uint32(26).fork()).ldelim();
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgRevokeNamespaceRoles();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.namespaceDenom = reader.string();
                    break;
                case 3:
                    message.addressRolesToRevoke.push(permissions_1.AddressRoles.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            namespaceDenom: isSet(object.namespaceDenom) ? String(object.namespaceDenom) : "",
            addressRolesToRevoke: Array.isArray(object === null || object === void 0 ? void 0 : object.addressRolesToRevoke)
                ? object.addressRolesToRevoke.map(function (e) { return permissions_1.AddressRoles.fromJSON(e); })
                : [],
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.namespaceDenom !== undefined && (obj.namespaceDenom = message.namespaceDenom);
        if (message.addressRolesToRevoke) {
            obj.addressRolesToRevoke = message.addressRolesToRevoke.map(function (e) { return e ? permissions_1.AddressRoles.toJSON(e) : undefined; });
        }
        else {
            obj.addressRolesToRevoke = [];
        }
        return obj;
    },
    create: function (base) {
        return exports.MsgRevokeNamespaceRoles.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b, _c;
        var message = createBaseMsgRevokeNamespaceRoles();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.namespaceDenom = (_b = object.namespaceDenom) !== null && _b !== void 0 ? _b : "";
        message.addressRolesToRevoke = ((_c = object.addressRolesToRevoke) === null || _c === void 0 ? void 0 : _c.map(function (e) { return permissions_1.AddressRoles.fromPartial(e); })) || [];
        return message;
    },
};
function createBaseMsgRevokeNamespaceRolesResponse() {
    return {};
}
exports.MsgRevokeNamespaceRolesResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgRevokeNamespaceRolesResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgRevokeNamespaceRolesResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgRevokeNamespaceRolesResponse();
        return message;
    },
};
function createBaseMsgClaimVoucher() {
    return { sender: "", denom: "" };
}
exports.MsgClaimVoucher = {
    encode: function (message, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        if (message.sender !== "") {
            writer.uint32(10).string(message.sender);
        }
        if (message.denom !== "") {
            writer.uint32(18).string(message.denom);
        }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgClaimVoucher();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.sender = reader.string();
                    break;
                case 2:
                    message.denom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (object) {
        return {
            sender: isSet(object.sender) ? String(object.sender) : "",
            denom: isSet(object.denom) ? String(object.denom) : "",
        };
    },
    toJSON: function (message) {
        var obj = {};
        message.sender !== undefined && (obj.sender = message.sender);
        message.denom !== undefined && (obj.denom = message.denom);
        return obj;
    },
    create: function (base) {
        return exports.MsgClaimVoucher.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (object) {
        var _a, _b;
        var message = createBaseMsgClaimVoucher();
        message.sender = (_a = object.sender) !== null && _a !== void 0 ? _a : "";
        message.denom = (_b = object.denom) !== null && _b !== void 0 ? _b : "";
        return message;
    },
};
function createBaseMsgClaimVoucherResponse() {
    return {};
}
exports.MsgClaimVoucherResponse = {
    encode: function (_, writer) {
        if (writer === void 0) { writer = minimal_js_1.default.Writer.create(); }
        return writer;
    },
    decode: function (input, length) {
        var reader = input instanceof minimal_js_1.default.Reader ? input : new minimal_js_1.default.Reader(input);
        var end = length === undefined ? reader.len : reader.pos + length;
        var message = createBaseMsgClaimVoucherResponse();
        while (reader.pos < end) {
            var tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON: function (_) {
        return {};
    },
    toJSON: function (_) {
        var obj = {};
        return obj;
    },
    create: function (base) {
        return exports.MsgClaimVoucherResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial: function (_) {
        var message = createBaseMsgClaimVoucherResponse();
        return message;
    },
};
var MsgClientImpl = /** @class */ (function () {
    function MsgClientImpl(rpc) {
        this.rpc = rpc;
        this.UpdateParams = this.UpdateParams.bind(this);
        this.CreateNamespace = this.CreateNamespace.bind(this);
        this.DeleteNamespace = this.DeleteNamespace.bind(this);
        this.UpdateNamespace = this.UpdateNamespace.bind(this);
        this.UpdateNamespaceRoles = this.UpdateNamespaceRoles.bind(this);
        this.RevokeNamespaceRoles = this.RevokeNamespaceRoles.bind(this);
        this.ClaimVoucher = this.ClaimVoucher.bind(this);
    }
    MsgClientImpl.prototype.UpdateParams = function (request, metadata) {
        return this.rpc.unary(exports.MsgUpdateParamsDesc, exports.MsgUpdateParams.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.CreateNamespace = function (request, metadata) {
        return this.rpc.unary(exports.MsgCreateNamespaceDesc, exports.MsgCreateNamespace.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.DeleteNamespace = function (request, metadata) {
        return this.rpc.unary(exports.MsgDeleteNamespaceDesc, exports.MsgDeleteNamespace.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.UpdateNamespace = function (request, metadata) {
        return this.rpc.unary(exports.MsgUpdateNamespaceDesc, exports.MsgUpdateNamespace.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.UpdateNamespaceRoles = function (request, metadata) {
        return this.rpc.unary(exports.MsgUpdateNamespaceRolesDesc, exports.MsgUpdateNamespaceRoles.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.RevokeNamespaceRoles = function (request, metadata) {
        return this.rpc.unary(exports.MsgRevokeNamespaceRolesDesc, exports.MsgRevokeNamespaceRoles.fromPartial(request), metadata);
    };
    MsgClientImpl.prototype.ClaimVoucher = function (request, metadata) {
        return this.rpc.unary(exports.MsgClaimVoucherDesc, exports.MsgClaimVoucher.fromPartial(request), metadata);
    };
    return MsgClientImpl;
}());
exports.MsgClientImpl = MsgClientImpl;
exports.MsgDesc = { serviceName: "injective.permissions.v1beta1.Msg" };
exports.MsgUpdateParamsDesc = {
    methodName: "UpdateParams",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgUpdateParams.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgUpdateParamsResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgCreateNamespaceDesc = {
    methodName: "CreateNamespace",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgCreateNamespace.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgCreateNamespaceResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgDeleteNamespaceDesc = {
    methodName: "DeleteNamespace",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgDeleteNamespace.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgDeleteNamespaceResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgUpdateNamespaceDesc = {
    methodName: "UpdateNamespace",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgUpdateNamespace.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgUpdateNamespaceResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgUpdateNamespaceRolesDesc = {
    methodName: "UpdateNamespaceRoles",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgUpdateNamespaceRoles.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgUpdateNamespaceRolesResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgRevokeNamespaceRolesDesc = {
    methodName: "RevokeNamespaceRoles",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgRevokeNamespaceRoles.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgRevokeNamespaceRolesResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
exports.MsgClaimVoucherDesc = {
    methodName: "ClaimVoucher",
    service: exports.MsgDesc,
    requestStream: false,
    responseStream: false,
    requestType: {
        serializeBinary: function () {
            return exports.MsgClaimVoucher.encode(this).finish();
        },
    },
    responseType: {
        deserializeBinary: function (data) {
            var value = exports.MsgClaimVoucherResponse.decode(data);
            return __assign(__assign({}, value), { toObject: function () {
                    return value;
                } });
        },
    },
};
var GrpcWebImpl = /** @class */ (function () {
    function GrpcWebImpl(host, options) {
        this.host = host;
        this.options = options;
    }
    GrpcWebImpl.prototype.unary = function (methodDesc, _request, metadata) {
        var _this = this;
        var _a;
        var request = __assign(__assign({}, _request), methodDesc.requestType);
        var maybeCombinedMetadata = metadata && this.options.metadata
            ? new browser_headers_1.BrowserHeaders(__assign(__assign({}, (_a = this.options) === null || _a === void 0 ? void 0 : _a.metadata.headersMap), metadata === null || metadata === void 0 ? void 0 : metadata.headersMap))
            : metadata || this.options.metadata;
        return new Promise(function (resolve, reject) {
            grpc_web_1.grpc.unary(methodDesc, {
                request: request,
                host: _this.host,
                metadata: maybeCombinedMetadata,
                transport: _this.options.transport,
                debug: _this.options.debug,
                onEnd: function (response) {
                    if (response.status === grpc_web_1.grpc.Code.OK) {
                        resolve(response.message.toObject());
                    }
                    else {
                        var err = new GrpcWebError(response.statusMessage, response.status, response.trailers);
                        reject(err);
                    }
                },
            });
        });
    };
    return GrpcWebImpl;
}());
exports.GrpcWebImpl = GrpcWebImpl;
var tsProtoGlobalThis = (function () {
    if (typeof globalThis !== "undefined") {
        return globalThis;
    }
    if (typeof self !== "undefined") {
        return self;
    }
    if (typeof window !== "undefined") {
        return window;
    }
    if (typeof global !== "undefined") {
        return global;
    }
    throw "Unable to locate global object";
})();
function isSet(value) {
    return value !== null && value !== undefined;
}
var GrpcWebError = /** @class */ (function (_super) {
    __extends(GrpcWebError, _super);
    function GrpcWebError(message, code, metadata) {
        var _this = _super.call(this, message) || this;
        _this.code = code;
        _this.metadata = metadata;
        return _this;
    }
    return GrpcWebError;
}(tsProtoGlobalThis.Error));
exports.GrpcWebError = GrpcWebError;
